boa/src/object/gcobject.rs

```diff
@@ -272,25 +288,23 @@ impl GcObject {
                     }
                 }
             } else {
-                let name = this.get_field("name", context)?.display().to_string();
+                let name = self.get(&"name".into(), context)?.display().to_string();
                 return context.throw_type_error(format!("{} is not a constructor", name));
             }
         } else {
             return context.throw_type_error("not a function");
         };
 
         match body {
-            FunctionBody::BuiltIn(function) => {
-                function(&this, args, context)?;
-                Ok(this)
-            }
+            FunctionBody::BuiltInConstructor(function) => function(&new_target, args, context),
             FunctionBody::Ordinary(body) => {
                 let _ = body.run(context);
 
                 // local_env gets dropped here, its no longer needed
                 let binding = context.realm_mut().environment.get_this_binding();
                 binding.map_err(|e| e.to_error(context))
             }
+            _ => unreachable!(),
```

Comment:
I would add some text in this `unreachable!()` call in case we ever mess up and end up here, in order to ease the debugging.